src/psi.rs

```diff
@@ -0,0 +1,276 @@
+        let k_i: Fq = permuted.pow(a);
+
+        // finally, hash x_i and k_i
+        let k_prime_i: Fq = hash_2(x_i, k_i);
+
+        capital_k.push(k_prime_i);
+    }
+
+    // step #6
+    // shuffle K
+    // https://rust-random.github.io/rand/rand/seq/trait.SliceRandom.html#example-4
+    let mut rng: ThreadRng = thread_rng();
+    capital_k.shuffle(&mut rng);
+
+    capital_k
+}
+
+/// Called by the receiver after receiving the output of `sender_2`.
+/// Step #7 in the paper.
+/// 
+/// # Arguments
+///
+/// * `capital_k` - Set outputted by `sender_2`
+/// * `m` - Field element received from the sender's first message (output of `sender_1`)
+/// * `b_i_array` - Set of random field elements generated by `receiver_1`
+/// * `set_y` - The receiver's private set
+/// 
+/// # Outputs
+///
+/// * `intersection` - The subset of the receiver's private set that only includes every element in the intersection
+pub fn receiver_2(capital_k: Vec<Fq>, m: Fq, b_i_array: Vec<BigInt<4>>, set_y: &Vec<u64>) -> Vec<u64> {
+    // step #7
+    let mut intersection: Vec<u64> = Vec::new();
+    for i in 0..set_y.len() {
+        // KA.key_2(b_i, m)
+        let key_2: Fq = m.pow(b_i_array[i]);
+
+        let hash: Fq = hash_2(set_y[i], key_2);
+
+        if capital_k.contains(&hash) {
+            intersection.push(set_y[i]);
+        }
+    }
+
+    intersection
+}
+
+/// Approximation for an ideal permutation.
+/// Uses AES GCM SIV (https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#AES-GCM-SIV)
+/// \Pi (and \Pi^{-1}) in the paper.
+/// 
+/// # Arguments
+///
+/// * `input` - A field element
+/// 
+/// # Outputs
+///
+/// * `permuted` - A field element
+fn pi(input: Fq) -> Fq {
+    // for the ideal permutation. because we need a simple fixed permutation, we don't need to change the key or nonce?
+    // TODO: those looooooong types are ugly
+    // TODO: these should be global variables
+    let key: aes::cipher::generic_array::GenericArray<u8, UInt<UInt<UInt<UInt<UInt<UTerm, B1>, B0>, B0>, B0>, B0>> = Aes128GcmSiv::generate_key(&mut OsRng);
+    let cipher: aes_gcm_siv::AesGcmSiv<aes::Aes128> = Aes128GcmSiv::new(&key);
+    let nonce: &aes::cipher::generic_array::GenericArray<u8, aes::cipher::typenum::UInt<aes::cipher::typenum::UInt<aes::cipher::typenum::UInt<aes::cipher::typenum::UInt<aes::cipher::typenum::UTerm, aes::cipher::typenum::B1>, aes::cipher::typenum::B1>, aes::cipher::typenum::B0>, aes::cipher::typenum::B0>> = Nonce::from_slice(b"unique nonce"); // 96-bits; unique per message
+
+    let input_string: String = std::format!("{input}");
+    // In AES, encryption and decryption are done by the same operation
+    // from: https://docs.rs/aes-gcm-siv/latest/aes_gcm_siv/#usage
+    // also: https://stackoverflow.com/questions/23850486/how-do-i-convert-a-string-into-a-vector-of-bytes-in-rust
+    // TODO: why does it give an error when I run `.decrypt`?
+    // TODO: is is_ok() okay (lol) or should I propagate the error with Result<T, E>?
```

Comment:
bubbling up the error is likely better so end user can know what went wrong

Reply:
would be help to understand the error, are these like IO errors or are these from bad user input stuff? 

general rule of thumb is to panic when the error is irrecoverable

Reply:
Looks like it's an OS error if `encrypt` fails. Here is the function:
```
    fn encrypt<'msg, 'aad>(
        &self,
        nonce: &Nonce<Self>,
        plaintext: impl Into<Payload<'msg, 'aad>>,
    ) -> Result<Vec<u8>> {
        let payload = plaintext.into();
        let mut buffer = Vec::with_capacity(payload.msg.len() + Self::TagSize::to_usize());
        buffer.extend_from_slice(payload.msg);
        self.encrypt_in_place(nonce, payload.aad, &mut buffer)?;
        Ok(buffer)
    }
```
`encrypt_in_place` "returns an error if the buffer has insufficient capacity to store the resulting ciphertext message."

Reply:
Keeping the code as is because that error is most likely irrecoverable.